runtimes/common/src/dip/merkle/v0/mod.rs

```diff
@@ -0,0 +1,333 @@
+	} = identity;
+	let mut trie = TrieHash::<LayoutV1<Runtime::Hashing>>::default();
+	let mut trie_builder = TrieDBMutBuilder::<LayoutV1<Runtime::Hashing>>::new(db, &mut trie).build();
+
+	// Authentication key.
+	let auth_leaves = get_auth_leaves(did_details)?;
+	// Attestation key, if present.
+	let att_leaves = get_att_leaves(did_details)?;
+	// Delegation key, if present.
+	let del_leaves = get_del_leaves(did_details)?;
+	// Key agreement keys.
+	let enc_leaves = get_enc_leaves(did_details)?;
+	// Linked accounts.
+	let linked_accounts = get_linked_account_leaves(linked_accounts);
+	// Web3name.
+	let web3_name = web3_name_details.as_ref().map(get_web3name_leaf::<Runtime>);
+
+	// Add all leaves to the proof builder.
+	let keys = auth_leaves
+		.chain(att_leaves)
+		.chain(del_leaves)
+		.chain(enc_leaves)
+		.map(RevealedDidMerkleProofLeaf::from);
+	let linked_accounts = linked_accounts.map(RevealedDidMerkleProofLeaf::from);
+	let web3_names = web3_name
+		.map(|n| vec![n])
+		.unwrap_or_default()
+		.into_iter()
+		.map(RevealedDidMerkleProofLeaf::from);
+
+	keys.chain(linked_accounts).chain(web3_names).try_for_each(|leaf| {
+		trie_builder
+			.insert(leaf.encoded_key().as_slice(), leaf.encoded_value().as_slice())
+			.map_err(|_| {
+				log::error!("Failed to insert leaf in the trie builder. Leaf: {:#?}", leaf);
+				DidMerkleProofError::Internal
+			})?;
+		Ok(())
+	})?;
+
+	trie_builder.commit();
+	Ok(trie_builder.root().to_owned())
+}
+
+/// Given the provided DID info, and a set of DID key IDs, account IDs and a
+/// web3name, generates a Merkle proof that reveals only the provided
+/// identity components. The function fails if no key or account with the
+/// specified ID can be found, or if a web3name is requested to be revealed
+/// in the proof but is not present in the provided identity details.
+pub(super) fn generate_proof<'a, Runtime, K, A, const MAX_LINKED_ACCOUNT: u32>(
+	identity: &LinkedDidInfoOf<Runtime, MAX_LINKED_ACCOUNT>,
+	key_ids: K,
+	should_include_web3_name: bool,
+	account_ids: A,
+) -> Result<CompleteMerkleProof<Runtime::Hash, DidMerkleProofOf<Runtime>>, DidMerkleProofError>
+where
+	Runtime: did::Config + pallet_did_lookup::Config + pallet_web3_names::Config,
+	K: Iterator<Item = &'a KeyIdOf<Runtime>>,
+	A: Iterator<Item = &'a LinkableAccountId>,
+{
+	let LinkedDidInfoOf {
+		did_details,
+		web3_name_details,
+		linked_accounts,
+	} = identity;
+
+	let mut db = MemoryDB::default();
+	let root = calculate_root_with_db(identity, &mut db)?;
+
+	let did_key_leaves_iter = key_ids.map(|key_id| -> Result<_, DidMerkleProofError> {
+		let key_details = did_details
+			.public_keys
+			.get(key_id)
+			.ok_or(DidMerkleProofError::KeyNotFound)?;
+		let key_relationships = {
+			if did_details.key_agreement_keys.contains(key_id) {
+				Ok(vec![DidKeyRelationship::Encryption])
+			} else {
+				let mut key_relationships = Vec::<DidKeyRelationship>::with_capacity(3);
+				if *key_id == did_details.authentication_key {
```

Comment:
This is the other bug that was found during testing. We used to exit after the first relationship was found, but we have to go over all the possible ones, and generate a new leaf for each.